### PR TITLE
Change Tokens topic

### LINKED ISSUE #1291: Consider adding topic that covers ChangeToken
`ChangeToken` is exposed in the configuration API surface but it is also a general purpose low level building block you can use by itself. 

Although using it to have logging automatically updated from configuration files can be very straightforward (e.g. it only requires setting a flag on the configuration provider), customers keep hitting some complexity when they try to use `ChangeToken` directly. Typically the first thing they hit is that it is a one time use rather than an event mechanism, so every time it fires you need to re-register to get a notification from the next `ChangeToken` if you need it.

Here is one example of a customer hitting this and the workaround: https://github.com/aspnet/Configuration/issues/432#issuecomment-221704063, but we had a few internal people hitting it during verification as well.

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #1291 
Addresses #4093 

[Internal review topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/primitives/change-tokens?branch=pr-en-us-4654)

### Notes
* To facilitate running the sample, I put it up here: https://github.com/guardrex/ChangeTokenSample *Teaser:* A little Poe for 🎃 Halloween ☠️ in there!
* I think we need to further develop the sentiment expressed in: "Callbacks are still guaranteed to fire, eventually." when describing `ActiveChangedCallbacks`.
* I'm a bit concerned about `Interlocked.Exchange` and `OnReload` with the description "trigger the configuration change token and create a new token." Is it accurate? Do we want to enhance the content there?
* I didn't get crazy with `CompositeChangeToken`. It just has a quick example with a few lines on behavior ... I shamelessly stole this from the tests. Do we want more there? ... or even less?
* Trying a new thing here: **The sample documents itself**. This vastly reduces the burden on the topic to document the sample's behaviors, allows the topic to focus on coding practices, reduces the images required for the topic (this topic requires no images), and allows the sample to change in ways that won't require a topic update in the future (e.g., screenshots).
* Trying out another new thing: Readers probably have a topic open in one window and the sample in another (i.e., their IDE). When they see code in the topic, they'll be naturally inclined to look for it in the sample (*I am!* :smile:). Many topics don't call out *where* the topic code example lives in the sample. I make it explicit throughout with good effect I think. For example: "The sample uses SHA1 file hashing (*Utilities/Utilities.cs*):"

cc/ @HaoK

### CODE REVIEW

**Comment by guardrex** [comment]
**WIP** Stand-by ... I'd like to touch the sample and object monitoring scenario. Should be quick ... I have an idea on what I'd like to do to enhance it.

**Comment by guardrex** [comment]
Very good ..  I'll work on it tonight.

**Comment by HaoK** [comment]
```
        public IDisposable RegisterChangeCallback(Action<object> callback, object state)
        {
            _callback = (messageChangeType) => callback(messageChangeType);
            return null;
        }
```

This is a bit weird, the callback is supposed to get the state instance you passed in via register, while this looks to ignore the state and read it from some external field...

**Comment by guardrex** [comment]
@HaoK Getting closer (I think).

See this version pertaining to file monitoring: https://github.com/guardrex/ChangeTokenSample/blob/master/Extensions/FileChangeToken.cs

Setup this way (*Startup.cs* `Configure`):
```csharp
PoemFileChangeToken = new FileChangeToken(config, env, fileName:"Spirits of the Dead", filePath:"poem.txt");
PoemFileChangeToken.StartMonitoring();
```

Can be enabled and disabled with the FileWatching page:
```csharp
public IActionResult OnPostStartMonitoring()
{
    Startup.PoemFileChangeToken.StartMonitoring();

    return RedirectToPage();
}

public IActionResult OnPostStopMonitoring()
{
    Startup.PoemFileChangeToken.StopMonitoring();

    return RedirectToPage();
}
```

One problem tho ... calling ...
```csharp
_monitoringObject.Dispose();
```
... isn't disabling the monitoring. It doesn't :boom:, but it doesn't work either. Callbacks continue to fire after calling `Dispose`.

I'm also getting an occasional file locking problem, so do I need to add retry logic with a short delay in my `ComputeHash` method (*Utilities/Utilities.cs*)?

**Comment by HaoK** [comment]
Try a basic test, before firing anything, can you call StartMonitoring and StopMonitoring to see if that disables them correctly?

**Comment by HaoK** [comment]
Re the locking, yeah you can just implement a simple retry like 3 times... and rethrow if that fails

**Comment by guardrex** [comment]
The start monitoring method is called in startup and works. The stop method fails.

I think its getting stuck on those double-callbacks that happen when watching files ... The dispose going on the prior token ... and the replacement token keeps working.

I'll work with more when I get home from the :muscle: gym and let u know what I find out.

**Comment by HaoK** [comment]
So let me take a step back here, is your FileChangeToken really a FileMonitor where you are expecting people to get notifications continuously when the file changes...  Or is it a FileChangeToken where you get one change, and then they need to get a new file change token to get any more changes.

ChangeTokens in general are not very friendly to use (as you are discovering :)...  The monitor behavior (which is closer to what `OnChange` provides) is usually a bit more intuitive to use...

**Comment by HaoK** [comment]
So the framework, Configuration and I guess caching expose the raw underlying change tokens for higher up concepts to plug in... Options hides this concept, it consumes change tokens but it exposes changes via IOptionsMonitor which has persistent change notifications via OnChange like semantics where you pass in an `Action` that gets called until you dispose it...

**Comment by guardrex** [comment]
> is your FileChangeToken really a FileMonitor where you are expecting people to get notifications continuously when the file changes

Yes ... but not so much "notifications," per se. It just allows you to run some other code and keep track of when a change has occurred based on the initial token returned by an `IFileProvider.Watch` implementation (i.e., `_env.ContentRootFileProvider.Watch(FilePath)`) and subsequent tokens created on callbacks. Perhaps there's no advantage demonstrated in this sample for file monitoring that couldn't be done just as well with a direct use of `FileSystemWatcher` because I'm not showing how to use these tokens with ... as you say ... higher-level concepts.

If that's the case, then why aren't we just showing the configuration change monitoring example in the sample and that's all we need to show? At least in that case, we're taking the token from config, using it to monitor for config changes where the dev has something they'd like to do, e.g., update a database, notify another system, etc. Showing the config-based example demos the value of tokens: Because config uses change tokens (and not some other approach, e.g., `FileSystemWatcher`), one can tap into it for a higher-level activity: run some code when config changes.

**Comment by guardrex** [comment]
@HaoK What do you think of @Tratcher's idea:

> Read a file into cache and add a change token that listens for file changes to remove the file from cache so it can be recreated.

I can give that a go. One ? tho: @Tratcher, do you envision that the file reload and new cache entry is handled by the PostEvictionDelegate? ... so all that the token action needs to do is trigger the cache item eviction? ... or should the token action do both ... evict and re-populate the cache with the file's contents?)

btw - I worry about this note for busy systems and doing something like this ...

> * Multiple requests can find the cached key value empty because the callback hasn't completed.
> * This can result in several threads repopulating the cached item.

... unless you think that if the cached key value *is empty* that it should read the file directly as a backstop. (That's kind'a scary, too ... file locks, for example).

The other thing that occurs is that we could leave this topic centered on the configuration example that we currently have. Now that I have an exponential back-off to deal with file locks, it should be safe and effective. It demos the concepts that you, @HaoK, described above (I'll add your sentiments to the topic) while providing a clear and useful implementation example. If the config example demos the concepts we need to show, we could hang our hat on it and move on.

**Comment by HaoK** [comment]
Yeah I think its probably better to split them up as you suggest (a config focused example, and you can have a separate topic which drills into the caching example/issues and just link them to each other).  I think people will likely being trying to use either config or caching when they are looking for an example anyways.

**Comment by HaoK** [comment]
I'm not opposed to you simplifying the example down to just basically illustrating how OptionsMonitor uses config to monitor changes...  That's more or less the canonical use case of how to build the higher level persistent monitoring using change tokens/OnChange.

**Comment by Tratcher** [comment]
As for the file eviction and reload, I only expect the token to handle the eviction. The consumer will handle the reload if the content is missing from the cache.

**Comment by guardrex** [comment]
@HaoK I put some hacks up in the repo sample to play with configuration monitoring alternatives :point_right: https://github.com/guardrex/ChangeTokenSample

Especially see:
* Simpler "simple" approach to monitoring config changes without state that doesn't require a token renewal :point_right: https://github.com/guardrex/ChangeTokenSample/blob/master/Startup.cs#L27-L30
* A service-based approach with switches+state a little more like OptionsMonitor :point_right: https://github.com/guardrex/ChangeTokenSample/blob/master/Extensions/ConfigurationMonitor1.cs
* A simplified service-based approach that just uses some props to control monitoring and also keeps state :point_right: https://github.com/guardrex/ChangeTokenSample/blob/master/Extensions/ConfigurationMonitor2.cs

Switches for turning the last two on and off are in the Index page code-behind :point_right: https://github.com/guardrex/ChangeTokenSample/blob/master/Pages/Index.cshtml.cs#L60-L90

After looking OptionsMonitor over and admiring its complexity, I don't see why if the topic is going to show/explain something *like OptionsMonitor* that it just doesn't *explain OptionsMonitor*.

Anyway ... these hacks were helpful. I was able to gain insight into OptionsMonitor mechanics and configure my prior example few different ways with switches and state.

**Comment by HaoK** [comment]
I'll take a look next week, but yeah the basic idea is the ChangeToken is the basic building block that you can build other things (like OptionsMonitor)

**Comment by guardrex** [comment]
@HaoK No rush at all ... I have *plenty* to keep me busy. :smile:

**Comment by guardrex** [comment]
@HaoK I'm circling back around to this one today and Friday. Let me try a section that explains change tokens using OptionsMonitor as the example. Then, I'll tack on my configuration monitoring bit in a section at the end.

**Comment by HaoK** [comment]
In the first "ChangeToken class" section, maybe you should combine the two overloads, just adding an blurb saying something like there's also an OnChange<TState> overload that takes an additional TState parameter that is passed into the token consumer Action.

In "monitoring for configuration changes" Maybe refer to the extension method "AddJsonFile" rather than the static class that adds the methods, i.e. "These files are configured using the AddJson extension method on ConfigurationBuilder via the reloadOnChange parameter"

Before you go into the filesystemwatcher, you should probably say that it uses the IFileProvider abstraction (https://github.com/aspnet/FileSystem/blob/dev/src/FS.Abstractions/IFileProvider.cs) to monitor files, which by default(https://github.com/aspnet/Configuration/blob/dev/src/Config.FileExtensions/FileConfigurationSource.cs#L82) uses PhysicalFileProvider which uses a filesystemwatcher to monitor changes.

Key point being you can plug in your own virtual file system as well (and hook up your own change monitoring)

**Comment by HaoK** [comment]
@ajcvickers the reload on change stuff has been a hotspot for config in the past, it's probably good to have a fresh (and less intimately familiar) set of eyes review this as well

**Comment by HaoK** [comment]
@guardrex The configuration monitor stuff looks fine, going with the simpler Monitor2 seems ok, was there any reason to prefer Monitor1?

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+`ChangeToken`'s [OnChange(Func\<IChangeToken>, Action)](/dotnet/api/microsoft.extensions.primitives.changetoken.onchange?view=aspnetcore-2.0#Microsoft_Extensions_Primitives_ChangeToken_OnChange_System_Func_Microsoft_Extensions_Primitives_IChangeToken__System_Action_) method registers an `Action` to call whenever the token changes:
+* `Func<IChangeToken>` produces the token.
+* `Action` is called when the token changes.
+
+`ChangeToken`'s [OnChange\<TState>(Func\<IChangeToken>, Action\<TState>, TState)](/dotnet/api/microsoft.extensions.primitives.changetoken.onchange?view=aspnetcore-2.0#Microsoft_Extensions_Primitives_ChangeToken_OnChange__1_System_Func_Microsoft_Extensions_Primitives_IChangeToken__System_Action___0____0_) method registers an `Action` to call whenever the token changes with state tracking:
+* `TState` is state for the token consumer `Action`.
+* `Func<IChangeToken>` produces the token.
+* `Action` is called when the token changes.
+
+## IChangeToken interface
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: Ordering? Move IChangeToken ahead of ChangeToken? IChangeToken is far more generally applicable.

**Comment by guardrex** [comment]
Thanks. I'll address all of that feedback :point_up: on the next commit.

@HaoK 

> The configuration monitor stuff looks fine, going with the simpler Monitor2 seems ok, was there any reason to prefer Monitor1?

No good reason really ... it's just a little closer to the structure of the OptionsMonitor code. It's over-engineered for what it does. I agree that the stripped-down version in *Startup.cs* (that runs 100% of the time) and Monitor2 (with the capacity to start and stop monitoring) are the best.

The stripped-down version is this one: 
```csharp
public void Configure(IApplicationBuilder app, IConfiguration config, IHostingEnvironment env)
{
    ChangeToken.OnChange(
        () => config.GetReloadToken(),
        (state) => InvokeChanged(state),
        env);
    ...
```
where `env` is the `IHostingEnvironment` and used like this in `InvokedChanged` ...
```csharp
private void InvokeChanged(IHostingEnvironment env)
{
    byte[] appsettingsHash = ComputeHash("appSettings.json");
    byte[] appsettingsEnvHash = 
        ComputeHash($"appSettings.{env.EnvironmentName}.json");

    if (!_appsettingsHash.SequenceEqual(appsettingsHash) || 
        !_appsettingsEnvHash.SequenceEqual(appsettingsEnvHash))
    {
        _appsettingsHash = appsettingsHash;
        _appsettingsEnvHash = appsettingsEnvHash;

        WriteConsole("Configuration changed (Simple)");
    }
}
```

WRT: Returning an `IDisposable` from `OnChange` that can be used to kill the token, I'll add that to the topic on the next commit.

@Tratcher 

> This seems really incomplete without some mention of using IFileProvider to produce change tokens and IMemoryCache to consume them (and not necessarily together).

:+1: I'll get that on the next commit.

... and if we don't explain out `OptionsMonitor` in detail, it doesn't mean that it can't be used as an example, too. I'll try to craft some language that talks about how `OptionsMonitor` uses change tokens without getting into the weeds on it.

**Comment by ajcvickers** [comment]
@HaoK Read through it and it makes sense to me. Is there something specific I should look at?

**Comment by guardrex** [comment]
@ajcvickers I have a new big commit on this PR coming soon ... either this afternoon or this evening. I'll ping you back here when the commit goes up.

**Comment by HaoK** [comment]
@ajcvickers nope, just wanted another set of eyes to make sure it made sense

**Comment by guardrex** [comment]
@ajcvickers @HaoK @Tratcher It's still a bit rough. I'll make some passes on it this weekend (grammar, style, etc.). However, try this latest version:

[Internal review topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/primitives/change-tokens?branch=pr-en-us-4654)

### Notes/questions
1. For the comment: "ActiveChangedCallbacks indicate if the token proactively raises callbacks. Callbacks are still guaranteed to fire, eventually." ... What does "callbacks are still guaranteed to fire, eventually" mean? It seems like we need to provide more explanation there, but idk what that means. (It came from the code/API comments.)
2. Is the content of the new *Example uses of Change Tokens* section (bullet list content) accurate? This is an attempt to address your, @Tratcher, comment about mentioning where/how they're used (e.g., `IFileProvider`, `IMemoryCache`). It might need a bit more :cow2: :bell: tho, so let me know if that's not enough.
3. Paragraphs 3 and 4 of the *Monitoring for configuration changes* section are now: Are those new paragraphs accurate?

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/primitives/change-tokens/sample/) ([how to download](xref:tutorials/index#how-to-download-a-sample))
+
+## IChangeToken interface
+
+The public interface [IChangeToken](/dotnet/api/microsoft.extensions.primitives.ichangetoken) propagates notifications that a change has occurred. `IChangeToken` resides in the [Microsoft.Extensions.Primitives](/dotnet/api/microsoft.extensions.primitives) namespace. For apps that don't use the [Microsoft.AspNetCore.All](https://www.nuget.org/packages/Microsoft.AspNetCore.All/) metapackage, reference the [Microsoft.Extensions.Primitives](https://www.nuget.org/packages/Microsoft.Extensions.Primitives/) NuGet package in the project file.
+
+The interface has two properties:
+
+* [ActiveChangedCallbacks](/dotnet/api/microsoft.extensions.primitives.ichangetoken.activechangecallbacks) indicate if the token proactively raises callbacks. Callbacks are still guaranteed to fire, eventually.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: `Callbacks are still guaranteed to fire, eventually.`? No they are not. If this is false then a callback will never be called, you need to poll `HasChanged` for changes. It's also possible a token will never be cancelled.

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+The interface has two properties:
+
+* [ActiveChangedCallbacks](/dotnet/api/microsoft.extensions.primitives.ichangetoken.activechangecallbacks) indicate if the token proactively raises callbacks. Callbacks are still guaranteed to fire, eventually. 
+* [HasChanged](/dotnet/api/microsoft.extensions.primitives.ichangetoken.haschanged) gets a value that indicates if a change has occurred.
+
+The interface has one method, [RegisterChangeCallback(Action\<Object>, Object)](/dotnet/api/microsoft.extensions.primitives.ichangetoken.registerchangecallback), which registers a callback that's invoked when the token has changed. `HasChanged` must be set before the callback is invoked.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: must -> will?

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+## Example uses of Change Tokens in ASP.NET Core
+
+Change Tokens are seen at work in promonent areas of ASP.NET Core monitoring changes to objects:
+
+* For monitoring changes to files, [IFileProvider](/dotnet/api/microsoft.extensions.fileproviders.ifileprovider)'s [Watch](/dotnet/api/microsoft.extensions.fileproviders.ifileprovider.watch) method creates an `IChangeToken` for the specified files or folder to watch.
+* For cache item expiration, the default [MemoryCache](/dotnet/api/microsoft.extensions.caching.memory.memorycache) implementation of [IMemoryCache](/dotnet/api/microsoft.extensions.caching.memory.imemorycache) offers a [Set](/dotnet/api/microsoft.extensions.caching.memory.cacheextensions.set?view=aspnetcore-2.0#Microsoft_Extensions_Caching_Memory_CacheExtensions_Set__1_Microsoft_Extensions_Caching_Memory_IMemoryCache_System_Object___0_Microsoft_Extensions_Primitives_IChangeToken_) overload that uses an `IChangeToken` as an `expirationToken` for the cached item.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: Drop the comment about the Set method, IChangeTokens are integrated much deeper than that. "IChangeToken's can be added to cache entries to trigger cache evictions on change."

**Comment by guardrex** [comment]
@Tratcher Thanks for reviewing. Do you think there's enough there for devs like @plaisted to go on when trying to implement change tokens for their purposes?

If not, then I can add some more :cow2: :bell: by implementing your earlier suggestion for cached file reload callbacks into caching. However if you and @HaoK think we have enough here, then I'll just address your feedback and move on to the next topic on my list.

**Comment by Tratcher** [comment]
I don't know who that is or what they were trying to do.

**Comment by guardrex** [comment]
Oh ... hold on ... I'll see if I can find that issue comment. You were helping him with change tokens IIRC.

**Comment by guardrex** [comment]
@Tratcher It's here :point_right: https://github.com/aspnet/Caching/issues/187#issuecomment-340815448

When you called out that 'the sample is being written now,' I'm just wondering that given what he's trying to accomplish if you think the sample today (and the content of this topic) is enough to help him ... or someone like him ... know enough about change tokens implementation to get their work done. Otherwise, it needs more work.

**Comment by Tratcher** [comment]
Ah. No, this is not enough yet. They need at least a minimal sample showing how to add IChangeTokens to a cache entry.

**Comment by guardrex** [comment]
I thought so. I have time today to work this forward.

File: aspnetcore/fundamentals/primitives/change-tokens.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/fundamentals/primitives/change-tokens.md]: @Tratcher I'll address this for the topic but note that it's in the comments:

https://github.com/aspnet/Common/blob/c9f4312474fa685f289f82d69288cd29b8a4f01a/src/Microsoft.Extensions.Primitives/IChangeToken.cs#L19

File: aspnetcore/fundamentals/primitives/change-tokens.md (hunk 3 above)
Comment (guardrex) [reply aspnetcore/fundamentals/primitives/change-tokens.md]: https://github.com/aspnet/Common/blob/c9f4312474fa685f289f82d69288cd29b8a4f01a/src/Microsoft.Extensions.Primitives/IChangeToken.cs#L25

**Comment by guardrex** [comment]
@Tratcher See this :point_right: https://github.com/guardrex/ChangeTokenSample ... particularly the `FileService` at ...

https://github.com/guardrex/ChangeTokenSample/blob/master/Services/FileService.cs

> Read a file into cache and add a change token that listens for file changes to remove the file from cache so it can be recreated.

I didn't follow you there on the *to remove the file from cache so it can be recreated* bit. That didn't make sense to me: I'd prefer to let sliding expiration remove file content from the cache and just use the token callback to update the cached content if the file is changed .... so that's what I did. I keep track of assigned tokens in a `List` to make sure I'm not creating more tokens than needed.

It works ... one prob tho ... I can't verify that the sliding expiration is working properly. If I let this app sit running beyond the sliding expiration window, it still loads the file content from cache. I'm not seeing an expected cache eviction after the window has passed. 😕

The callback works. Updating the file triggers the token's callback, and the file contents are updated in the cache.

**Comment by Tratcher** [comment]
Where is the GetFileContent method? https://github.com/guardrex/ChangeTokenSample/blob/e0a526ac53f2b21548ca4b0a66b96a3ef48e89bb/Services/FileService.cs#L42

This pattern doesn't work. You have it set up so that if the user asks for a file and it's not in the cache then it's added with a 5 minute sliding expiration. That's fine. However every time the file changes it's added to the cache again with no expiration, whether or not the consumer ever asks for it again. So if the user asks for a file only once, and then the file is changed 20 times, it gets added back to the cache 20 times. The value in the cache should only be reloaded if the consumer still wants it.

Try this:
```
            var changeToken = _fileProvider.Watch(fileName);
            fileContent = await GetFileContent(filePath);

            if (fileContent != null)
            {
                var cacheEntryOptions = new MemoryCacheEntryOptions()
                    .SetSlidingExpiration(TimeSpan.FromMinutes(5))
                    .AddExpirationToken(changeToken));

                // Put the file contents into the cache.
                _cache.Set(filePath, fileContent, cacheEntryOptions);

                return fileContent;
            }

            return string.Empty;
```

**Comment by guardrex** [comment]
@Tratcher Thanks for setting me straight.

`GetFileContent` is a utility method:

https://github.com/guardrex/ChangeTokenSample/blob/e0a526ac53f2b21548ca4b0a66b96a3ef48e89bb/Utilities/Utilities.cs#L56

I should've paid more attention to Intellisense (or the ref source itself). I wasn't aware of `AddExpirationToken` simply because it wasn't in the in-memory caching topic at https://docs.microsoft.com/aspnet/core/performance/caching/memory.

I updated the sample and the topic. I fixed up the cross-ref links in the caching group of topics and link over to this new topic.

I'm going to hit the :muscle: gym for a bit then give the topic and sample one more pass before taking it off **WIP**. I'll ping u back.

**Comment by guardrex** [comment]
@Tratcher Ok ... I think it's ready for a look. Given that it's :turkey: week, I'm OOF this week on vacation and don't expect this is a good week for feedback anyway. However if you want to review, I'll react to feedback on Monday, 11/27. *Happy T-Bird Day!* and I'll catch you after I gain a pound (or two! lol) 🍗.

**Comment by Rick-Anderson** [comment]
@Tratcher can you give this the final sign-off?

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/primitives/change-tokens/sample/) ([how to download](xref:tutorials/index#how-to-download-a-sample))
+
+## IChangeToken interface
+
+The public interface [IChangeToken](/dotnet/api/microsoft.extensions.primitives.ichangetoken) propagates notifications that a change has occurred. `IChangeToken` resides in the [Microsoft.Extensions.Primitives](/dotnet/api/microsoft.extensions.primitives) namespace. For apps that don't use the [Microsoft.AspNetCore.All](https://www.nuget.org/packages/Microsoft.AspNetCore.All/) metapackage, reference the [Microsoft.Extensions.Primitives](https://www.nuget.org/packages/Microsoft.Extensions.Primitives/) NuGet package in the project file.
+
+The interface has two properties:
+
+* [ActiveChangedCallbacks](/dotnet/api/microsoft.extensions.primitives.ichangetoken.activechangecallbacks) indicate if the token proactively raises callbacks. If `ActiveChangedCallbacks` is set to `false`, a callback is never called, and the app must poll `HasChanged` for changes. It's also possible a token will never be cancelled when `ActiveChangedCallbacks` is set to `false`.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: "It's also possible for a token to never be cancelled if no changes occur or the underlying change listener is cleaned up." This is unrelated to ActiveChangeCallbacks

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+[!code-cshtml[Main](change-tokens/sample/Pages/Index.cshtml?range=35)]
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet2)]
+
+When the user triggers the `OnPostStartMonitoring` method, monitoring is enabled and the current state is cleared. When the user triggers the `OnPostStopMonitoring` method, monitoring is disabled and the state is set to reflect that monitoring is not occurring.
+
+## Monitoring cached file changes
+
+File content is cached in-memory using [IMemoryCache](/dotnet/api/microsoft.extensions.caching.memory.imemorycache). In-memory caching is described in the [In-memory caching](xref:performance/caching/memory) topic. Without taking additional steps, *stale* (outdated) data is returned from a cache if the source data has changed.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: File content can be cached

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+For example, ignoring the status of a source file for cached file data on a [sliding expiration](/dotnet/api/microsoft.extensions.caching.memory.memorycacheentryoptions.slidingexpiration) period leads to stale cache data when the file is changed but the caching system isn't notified. Each request for the data renews the sliding expiration period, but the file is never reloaded into the cache. Any app features that use the file's cached content are subject to possibly receiving stale content.
+
+Using change tokens in a file caching scenario avoids the possibility of recovering stale file content from the cache. The sample app demonstrates an implementation of the approach.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: recovering -> getting

**Comment by guardrex** [comment]
@Tratcher Thanks. @scottaddie @Rick-Anderson Ready to launch. :rocket:

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+## IChangeToken interface
+
+The public interface [IChangeToken](/dotnet/api/microsoft.extensions.primitives.ichangetoken) propagates notifications that a change has occurred. `IChangeToken` resides in the [Microsoft.Extensions.Primitives](/dotnet/api/microsoft.extensions.primitives) namespace. For apps that don't use the [Microsoft.AspNetCore.All](https://www.nuget.org/packages/Microsoft.AspNetCore.All/) metapackage, reference the [Microsoft.Extensions.Primitives](https://www.nuget.org/packages/Microsoft.Extensions.Primitives/) NuGet package in the project file.
+
+The interface has two properties:
+
+* [ActiveChangedCallbacks](/dotnet/api/microsoft.extensions.primitives.ichangetoken.activechangecallbacks) indicate if the token proactively raises callbacks. If `ActiveChangedCallbacks` is set to `false`, a callback is never called, and the app must poll `HasChanged` for changes. It's also possible for a token to never be cancelled if no changes occur or the underlying change listener is cleaned up.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: What does cleaned up mean?

Comment (guardrex) [reply aspnetcore/fundamentals/primitives/change-tokens.md]: I took it to mean calling `Dispose` on the token. It should be more specific.

Comment (Tratcher) [reply aspnetcore/fundamentals/primitives/change-tokens.md]: It's provider specific. Disposing of the underlying CancellationToken is one example.

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+An instance `ConfigurationMonitor` is registered as a service in `ConfigureServices` of *Startup.cs*:
+
+[!code-csharp[Main](change-tokens/sample/Startup.cs?name=snippet1)]
+
+The Index page offers the user control over configuration monitoring. The instance of `IConfigurationMonitor` is injected into the `IndexModel`:
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet1)]
+
+The app responds on form submits when the user selects buttons in the interface to enable and disable monitoring:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: A button enables/disables monitoring.

No need to provide all those other details.

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+[!code-cshtml[Main](change-tokens/sample/Pages/Index.cshtml?range=35)]
+
+[!code-csharp[Main](change-tokens/sample/Pages/Index.cshtml.cs?name=snippet2)]
+
+When the user triggers the `OnPostStartMonitoring` method, monitoring is enabled and the current state is cleared. When the user triggers the `OnPostStopMonitoring` method, monitoring is disabled and the state is set to reflect that monitoring is not occurring.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: +When `OnPostStartMonitoring`  is triggers , monitoring is enabled and the current state is cleared.
ditto for the next method. Less is more better.

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+## Monitoring cached file changes
+
+File content can be cached in-memory using [IMemoryCache](/dotnet/api/microsoft.extensions.caching.memory.imemorycache). In-memory caching is described in the [In-memory caching](xref:performance/caching/memory) topic. Without taking additional steps, *stale* (outdated) data is returned from a cache if the source data has changed.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: >Without taking additional steps,

I get it - unless you configure caching, caching doesn't work.

So are you presenting an alternate approach to caching?

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+Using change tokens in a file caching scenario avoids the possibility of getting stale file content from the cache. The sample app demonstrates an implementation of the approach.
+
+The sample establishes a static method, `GetFileContent`, to return file content. The method uses a retry algorithm with exponential back-off to cover cases where a file lock is temporarily preventing a file from being read (*Utilities/Utilities.cs*):
+
+[!code-csharp[Main](change-tokens/sample/Utilities/Utilities.cs?name=snippet2)]
+
+A `FileService` is created to handle cached file lookups. The `GetFileContent` method call of the service attempts to obtain file content from the in-memory cache and return it to the caller (*Services/FileService.cs*).
+
+If cached content isn't found using the cache key, which is the file path in this example, the following actions are taken:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: ~which is the file path in this example~
Avoid the temptation to provide too many details.

File: aspnetcore/fundamentals/primitives/change-tokens.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+1. The file content is obtained using `GetFileContent`.
+1. A change token is obtained from the file provider with [IFileProviders.Watch](/dotnet/api/microsoft.extensions.fileproviders.ifileprovider.watch) whose callback is triggered when the file is modified.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/primitives/change-tokens.md]: >whose callback is triggered when the file is modified
sentence too long. Split that up.

**Comment by guardrex** [comment]
@Rick-Anderson I addressed everything except for this pesky line ...

> It's also possible for a token to never be cancelled if no changes occur or the underlying change listener is cleaned up.

**Comment by Rick-Anderson** [comment]
@guardrex  I like "It's also possible for a token to never be cancelled if no changes occur or the underlying change listener is disposed or disabled."

add that and I'll merge

**Comment by guardrex** [comment]
😅 ... *It's Miller Time*:tm: on this one. :beer:

"It's Miller Time" is a registered trademark of Miller Brewing Co. :grinning:
